File: packages/lutrachat_backend/lutrachat_backend_common/lib/src/service/token.dart
import 'dart:typed_data';

import '../../gen/proto/lutracorp/lutrachat/common/service/token/v1/data.pb.dart';

/// A service for signing and verifying tokens.
abstract interface class TokenService {
  /// Decodes the token into data.
  TokenData decode(String token);

  /// Signs the token with a secret key.
  Future<String> sign(
    Uint8List payload,
    Uint8List secretKey, [
    DateTime? timestamp,
  ]);

  /// Checks the validity of the token.
  Future<bool> verify(String token, Uint8List secretKey);
}

File: internal/pkg/token/token.go
package token

import (
	tokpb "github.com/lutracorp/lutrachat/api/protocol/pkg/token/v1"
)

var handler *Handler

// Initialize creates handler used to manipulate tokens.
func Initialize(cfg *Config) error {
	var err error
	handler, err = NewHandler(cfg)

	return err
}

// Sign signs payload with secret and returns TokenData.
func Sign(payload []byte, secret []byte) (*tokpb.TokenData, error) {
	return handler.Sign(payload, secret)
}

// Verify verifies signed token data for validity.
func Verify(data *tokpb.TokenData, secret []byte) (bool, error) {
	return handler.Verify(data, secret)
}
